modifiers/godot/voxel_modifier_gd.cpp

```diff
@@ -115,16 +85,55 @@ void VoxelModifier::_notification(int p_what) {
 				_modifier_id = 0;
 			}
 		} break;
+		case Node3D::NOTIFICATION_TRANSFORM_CHANGED: {
+			if (_volume == nullptr) {
+				Node *parent = get_parent();
+				VoxelLodTerrain *volume = Object::cast_to<VoxelLodTerrain>(parent);
+				// Using get_parent_node_3d when terrain is immediate parent
+				// would be a breaking change as it returns null if node is toplevel
+				if (volume == nullptr) {
+					Node3D *grandparent = get_parent_node_3d();
+					volume = Object::cast_to<VoxelLodTerrain>(grandparent);
+					while (grandparent != nullptr && volume == nullptr) {				
+						grandparent = grandparent->get_parent_node_3d();
+						volume = Object::cast_to<VoxelLodTerrain>(grandparent);
+					}
+				}
+				_volume = volume;
+			
+
+				if (_volume != nullptr) {
+					VoxelData &voxel_data = _volume->get_storage();
+					VoxelModifierStack &modifiers = voxel_data.get_modifiers();
+					const uint32_t id = modifiers.allocate_id();
+					zylann::voxel::VoxelModifier *modifier = create(modifiers, id);
+
+					if (modifier->is_sdf()) {
+						zylann::voxel::VoxelModifierSdf *sdf_modifier =
+								static_cast<zylann::voxel::VoxelModifierSdf *>(modifier);
+						sdf_modifier->set_operation(to_op(_operation));
+						sdf_modifier->set_smoothness(_smoothness);
+					}
+					// This transform is wrong, but will be updated with NOTIFICATION_TRANSFORM_CHANGED
+					// After node enters the tree
+					modifier->set_transform(get_transform());
```

Comment:
In case the modifier already is registered in the terrain, then you can get a valid transform, because modifier transforms are relative to the terrain. It doesn't matter whether they are in the tree or not. And once that's done, it should not have to be set again, until the modifier or any of its parents below the terrain are changed.
So I'm not sure you should set a wrong transform here...

Reply:
For the very last part, do you mean the updates should be batched initially somehow?

Reply:
What I mean is that the fact the terrain gets added to the scene tree may already trigger updates on its own, but if modifiers also trigger individual updates on top of that, it could mean unnecessary updates depending on the timing. Although I'm not certain that this really matters, it might already have been a problem before your PR.